### PR TITLE
Hide archived labels when filtering by labels on the issue list

### PR DESCRIPTION
**Pull request by puni9869**
Followup  https://github.com/go-gitea/gitea/pull/26820
## Archived labels UI for issue filter and issue filter actions for issues/pull request pages.

Changed:
* Enhanced the Issue filter and Issue filter actions  UI page to seamlessly incorporate a list of archived labels.
* Pagination functionality is same as before. If archived label checkbox is checked then we are adding a query string`archived=true` in the url to save the state of page.
* Issue filter actions menu is separated into different template.
* Adding the archived flag in issue url labels.
* Pull Request page is also work the same.

Outsourced:
* Defer the implementation of specialized handling for archived labels to upcoming pull requests. This step will be undertaken subsequent to the successful merge of this pull request.

Screenshots
### Issue page
image

### Issue page with label filter on archived label checkbox when not checked --> No archived label is there in list
image

### Issue page with label filter on archived label checkbox when checked --> Show archived label in the list.
image

### Issue page with label filter on issue action menu on archived label checkbox when checked --> Show archived label in the list.
image

### Applied the archived=true in Issue labels when archived checkbox is checked.
image

---

Part of https://github.com/go-gitea/gitea/issues/25237

### CODE REVIEW

**Comment by silverwind** [comment]
PR is empty.

**Comment by puni9869** [comment]
Yes I am working on it...

File: web_src/js/features/repo-issue-list.js [hunk 1]
```diff
@@ -185,9 +185,20 @@ async function initIssuePinSort() {
...
+function initArchivedLabelFilter() {
+  const archivedLabelEl = document.querySelector('#archived-filter-checkbox');
+  if (archivedLabelEl) {
+    archivedLabelEl.addEventListener('change', () => {
+      const url = archivedLabelEl.getAttribute('data-url');
+      window.location = (archivedLabelEl.checked) ? url : url.replace(/&archived=true/, '');
+    });
+  }
+}
+
```
Comment (puni9869) [inline web_src/js/features/repo-issue-list.js:197]: Alternate implementation
```js
function initArchivedLabelFilter() {
  const archivedLabelEl = document.querySelector('#archived-filter-checkbox');

  if (archivedLabelEl) {
    archivedLabelEl.addEventListener('change', () => {
      const url = new URL(window.location.href);
      const isChecked = archivedLabelEl.checked;

      if (isChecked) {
        url.searchParams.delete('archived');
      } else {
        url.searchParams.set('archived', 'true');
      }

      window.location.href = url.toString();
    });
  }
}
```

**Comment by puni9869** [comment]
> PR is empty.

Ready for review

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1512,7 +1512,9 @@ issues.label_description = Description
 issues.label_color = Color
 issues.label_exclusive = Exclusive
 issues.label_archive = Archive Label
-issues.label_archive_tooltip= Archived labels are excluded from the label search when applying labels to an issue. Existing labels on issues remain unaffected, allowing you to retire obsolete labels without losing information.
+issues.label_archived  = Archived
+issues.label_archived_filter = Archived Labels
```
Comment (delvh) [inline options/locale/locale_en-US.ini]:
```suggestion
issues.label_archived_filter = Show archived labels
```

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1512,7 +1512,9 @@ issues.label_description = Description
...
+issues.label_archive_tooltip = Archived labels are excluded from the label search when applying labels to an issue. Existing labels on issues remain unaffected, allowing you to retire obsolete labels without losing information.
```
Comment (delvh) [inline options/locale/locale_en-US.ini]: The second sentence is unnecessary in this context 
```suggestion
issues.label_archive_tooltip = Archived labels are excluded by default from the suggestions when searching by label.
```

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -1512,7 +1512,9 @@ issues.label_description = Description
 issues.label_color = Color
 issues.label_exclusive = Exclusive
 issues.label_archive = Archive Label
-issues.label_archive_tooltip= Archived labels are excluded from the label search when applying labels to an issue. Existing labels on issues remain unaffected, allowing you to retire obsolete labels without losing information.
+issues.label_archived  = Archived
```
Comment (delvh) [inline options/locale/locale_en-US.ini]: Don't we already have this translation key?
Can we perhaps merge the already existing `<no section>.archived` into this translation?

File: routers/web/repo/issue.go [hunk 5]
```diff
@@ -356,6 +356,17 @@ func issues(ctx *context.Context, milestoneID, projectID int64, isPullOption uti
 		labels = append(labels, orgLabels...)
 	}
 
+	if !ctx.FormBool("archived") {
```
Comment (delvh) [inline routers/web/repo/issue.go]: Hmm… That is now a far more complicated approach than I imagined.
I think this is one of the features that is better implemented in the frontend:
Simply give **all** labels to the frontend as before, add a `data-is-archived-label="true"` if the label is archived, and then run JS over it that is executed once on page load and then whenever the checkbox changes its state that toggles the `gt-hidden` class for all elements in `document.querySelectorAll('[data-is-archived=true]')` (or whatever the exact syntax is) depending on if the checkbox is false or not.
Then there is no need to torment the backend with requests or to do some URL trickery anymore.

Comment (puni9869) [reply routers/web/repo/issue.go]: we can discuss on this approach.

Comment (puni9869) [reply routers/web/repo/issue.go]: Done

Comment (puni9869) [reply routers/web/repo/issue.go]: As discussed I have moved this approach into frontend side.

* Updates
Label issue action updated --> Archived hint is in right end.
image

File: web_src/js/features/repo-issue-list.js [hunk 6]
```diff
@@ -185,9 +185,27 @@ async function initIssuePinSort() {
...
+  archivedLabelEl.addEventListener('change', () => {
+    const url = archivedLabelEl.getAttribute('data-url');
+    window.location = (archivedLabelEl.checked) ? url : url.replace(/&archived=true/, '');
+  });
+  const archivedLabels = $('[data-is-archived]');
+  if (!archivedLabelEl.checked) {
+    for (const label of archivedLabels) {
+      toggleElem(label, archivedLabelEl.checked);
+    }
+  }
```
Comment (delvh) [inline web_src/js/features/repo-issue-list.js]:
```suggestion
  for (const label of archivedLabels) {
    toggleElem(label, archivedLabelEl.checked);
  }
  // TODO: Also show archived labels as soon as the checkbox changes its value. Depends on the page not being reloaded first.
```

File: templates/repo/issue/filter_actions.tmpl [hunk 7]
```diff
@@ -0,0 +1,130 @@
+<div class="ui secondary filter menu">
+	{{if not .Repository.IsArchived}}
+		<!-- Action Button -->
+		{{if .IsShowClosed}}
+			<button class="ui primary basic button issue-action" data-action="open" data-url="{{$.RepoLink}}/issues/status">{{.locale.Tr "repo.issues.action_open"}}</button>
```
Comment (delvh) [inline templates/repo/issue/filter_actions.tmpl]: Please convert all `.locale.Tr` to `ctx.Locale.Tr`.

**Comment by delvh** [comment]
Found a bug:
Given this state
grafik
(URL `/owner/repo/issues?q=&type=all&sort=&state=open&labels=5&milestone=0&project=0&assignee=0&poster=0&archived=true`),
if you click on the `Show archived labels` checkbox, the query loses its `labels=5` parameter:
The result is the URL `/owner/repo/issues?q=&type=all&sort=&state=open&milestone=0&project=0&assignee=0&poster=0`

**Comment by delvh** [comment]
Apart from these three things LGTM.

File: web_src/js/features/repo-issue-list.js [hunk 8]
```diff
@@ -185,9 +185,58 @@ async function initIssuePinSort() {
...
+//       toggleElem(label, archivedLabelEl.checked);
+//     }
+//   };
+//   archivedElToggle();
+//   archivedLabelEl.addEventListener('change', () => {
+//     archivedElToggle();
+//     const url = archivedLabelEl.getAttribute('data-url');
+//     window.location = (archivedLabelEl.checked) ? url : url.replace(/&archived=true/, '');
+//   });
+// }
```
Comment (delvh) [inline web_src/js/features/repo-issue-list.js]:
```suggestion
```

**Comment by puni9869** [comment]
I am working on its last step. Will push rest of the code tomorrow..

File: web_src/js/features/repo-issue-list.js (hunk 8 above)
Comment (puni9869) [reply web_src/js/features/repo-issue-list.js]: Done

File: web_src/js/features/repo-issue-list.js [hunk 9]
```diff
@@ -185,51 +185,24 @@ async function initIssuePinSort() {
...
 function initArchivedLabelFilter() {
   const archivedLabelEl = document.querySelector('#archived-filter-checkbox');
   if (!archivedLabelEl) {
     return;
   }
 
-  const archivedLabels = document.querySelectorAll('[data-is-archived]');
-  const toggleArchivedLabels = () => {
-    const isChecked = archivedLabelEl.checked;
+  const archivedLabels = $('[data-is-archived]');
```
Comment (delvh) [inline web_src/js/features/repo-issue-list.js]: Why suddenly jquery for this single call that was previously not JQuery?

File: web_src/js/features/repo-issue-list.js [hunk 10]
```diff
@@ -185,51 +185,24 @@ async function initIssuePinSort() {
...
-  toggleArchivedLabels();
-
-  // Event listeners
+  archivedElToggle();
   archivedLabelEl.addEventListener('change', () => {
-    toggleArchivedLabels();
-    updateURL();
+    archivedElToggle();
+    const url = archivedLabelEl.getAttribute('data-url');
+    window.location = (archivedLabelEl.checked) ? url : url.replace(/&archived=true/, '');
```
Comment (delvh) [inline web_src/js/features/repo-issue-list.js]: This code will break as soon as the `archived` query param is at the beginning, btw.

File: routers/web/repo/issue.go [hunk 11]
```diff
@@ -413,6 +413,22 @@ func issues(ctx *context.Context, milestoneID, projectID int64, isPullOption uti
...
+				if labelId == l.ID && !l.IsArchived() {
+					tempLabelIds = append(tempLabelIds, labelId)
+				}
+			}
+		}
+		if len(tempLabelIds) > 0 {
+			labelIDs = tempLabelIds
+			selectLabels = strings.Join(base.Int64sToStrings(labelIDs), ",")
+		}
+	}
```
Comment (delvh) [inline routers/web/repo/issue.go]: Wait, why is this suddenly a backend task again?
I thought we discussed that this is a frontend-only task (where we simply pass through the query param)?

File: web_src/js/features/repo-issue-list.js (hunk 10 above)
Comment (puni9869) [reply web_src/js/features/repo-issue-list.js]: Thanks for pointing out this.

File: routers/web/repo/issue.go (hunk 11 above)
Comment (puni9869) [reply routers/web/repo/issue.go]: This is not needed.

File: web_src/js/features/repo-issue-list.js (hunk 9 above)
Comment (puni9869) [reply web_src/js/features/repo-issue-list.js]: yeah, we can change this.

File: templates/repo/issue/filters.tmpl [hunk 12]
```diff
@@ -41,7 +41,16 @@
 							<div class="divider"></div>
 						{{end}}
 						{{$previousExclusiveScope = $exclusiveScope}}
-						<a class="item label-filter-item gt-df gt-ac" {{if .IsArchived}}data-is-archived{{end}} href="{{$.Link}}?q={{$.Keyword}}&type={{$.ViewType}}&sort={{$.SortType}}&state={{$.State}}&labels={{.QueryString}}&milestone={{$.MilestoneID}}&project={{$.ProjectID}}&assignee={{$.AssigneeID}}&poster={{$.PosterID}}{{if $.ShowArchivedLabels}}&archived=true{{end}}" data-label-id="{{.ID}}">{{if .IsExcluded}}{{svg "octicon-circle-slash"}}{{else if .IsSelected}}{{if $exclusiveScope}}{{svg "octicon-dot-fill"}}{{else}}{{svg "octicon-check"}}{{end}}{{end}}
+						<a class="item label-filter-item gt-df gt-ac" {{if .IsArchived}}data-is-archived{{end}} href="{{$.Link}}?q={{$.Keyword}}&type={{$.ViewType}}&sort={{$.SortType}}&state={{$.State}}&labels={{.QueryString}}&milestone={{$.MilestoneID}}&project={{$.ProjectID}}&assignee={{$.AssigneeID}}&poster={{$.PosterID}}{{if $.ShowArchivedLabels}}&archived=true{{end}}" data-is-checked data-label-id="{{.ID}}">
```
Comment (delvh) [inline templates/repo/issue/filters.tmpl]:
```suggestion
						<a class="item label-filter-item gt-df gt-ac" {{if .IsArchived}}data-is-archived{{end}} href="{{$.Link}}?q={{$.Keyword}}&type={{$.ViewType}}&sort={{$.SortType}}&state={{$.State}}&labels={{.QueryString}}&milestone={{$.MilestoneID}}&project={{$.ProjectID}}&assignee={{$.AssigneeID}}&poster={{$.PosterID}}{{if $.ShowArchivedLabels}}&archived=true{{end}}" data-label-id="{{.ID}}">
```

File: web_src/js/features/repo-issue-list.js [hunk 13]
```diff
@@ -191,18 +191,31 @@ function initArchivedLabelFilter() {
...
+    .split(',')
+    .map((id) => id < 0 ? `${~id + 1}` : id); // selectedLabels contains -ve ids, which are excluded so convert any -ve value id to +ve
+
   const archivedElToggle = () => {
     for (const label of archivedLabels) {
-      toggleElem(label, archivedLabelEl.checked);
+      const id = label.getAttribute('data-label-id');
+      if (selectedLabels.length && !selectedLabels.includes(id)) {
+        toggleElem(label, archivedLabelEl.checked);
+      }
```
Comment (delvh) [inline web_src/js/features/repo-issue-list.js]:
```suggestion
       toggleElem(label, archivedLabelEl.checked || selectedLabels.includes(id));
```

File: templates/repo/issue/filters.tmpl [hunk 14]
```diff
@@ -18,18 +18,39 @@
 						<i class="icon">{{svg "octicon-search" 16}}</i>
 						<input type="text" placeholder="{{ctx.Locale.Tr "repo.issues.filter_label"}}">
 					</div>
+					<div class="ui checkbox compact archived-label-filter">
+						<input name="archived" type="checkbox"
+							id="archived-filter-checkbox"
+							{{if .ShowArchivedLabels}}checked{{end}}
+						>
+						<label>
```
Comment (delvh) [inline templates/repo/issue/filters.tmpl]:
```suggestion
						<label for="archived-filter-checkbox">
```

File: templates/repo/issue/filters.tmpl [hunk 15]
```diff
@@ -18,18 +18,39 @@
 						<i class="icon">{{svg "octicon-search" 16}}</i>
 						<input type="text" placeholder="{{ctx.Locale.Tr "repo.issues.filter_label"}}">
 					</div>
+					<div class="ui checkbox compact archived-label-filter">
+						<input name="archived" type="checkbox"
```
Comment (delvh) [inline templates/repo/issue/filters.tmpl:22]: Hmm…
This checkbox is currently completely unaccessible.
I see, here's another candidate for where we need to improve the accessibility offered by Fomantic…

EDIT: Okay, surprisingly, there is one weird workaround to make it "accessible": Click on the label name field, then you can tab into the checkbox…
But I would not call that "accessible".

**Review by delvh — APPROVED** [review]

**Comment by puni9869** [comment]
Need one approval on this and more reviews.

File: templates/repo/issue/filters.tmpl (hunk 15 above)
Comment (puni9869) [reply templates/repo/issue/filters.tmpl:22]: the same

**Comment by delvh** [comment]
This PR should be backported as the functionality is simply missing in 1.21 otherwise.

**Review by lunny — APPROVED** [review]

**Review by denyskon — APPROVED** [review]

**Comment by puni9869** [comment]
Backport 1.21

**Pull request merged**